### PR TITLE
Improved fixture and test naming scheme

### PR DESCRIPTION
**Pull request by vkarak**
This is along my proposal in #1794.

### CODE REVIEW

File: reframe/core/pipeline.py [hunk 1]
```diff
@@ -895,7 +895,7 @@ def __init_subclass__(cls, *, special=False, pin_prefix=False,
     def __rfm_init__(self, *args, prefix=None, **kwargs):
         if not hasattr(self, 'name'):
             try:
-                self.name = type(self).fullname(self.variant_num)
+                self.name = type(self).basename(self.variant_num)
```
Comment (victorusu) [inline reframe/core/pipeline.py:898]: I am see the use of `basename` using `variant_num`... When are we using it with `compress_params` and `prefix`?

File: reframe/core/meta.py [hunk 2]
```diff
@@ -784,3 +787,32 @@ def fullname(cls, variant_num=None, *, compress_params=False):
                                    if fs[k].cls.num_variants > 1) + '^'
 
         return name
+
+    def basename(cls, variant_num=None, *,
+                 compress_params=False, prefix=''):
```
Comment (victorusu) [inline reframe/core/meta.py:792]: I think that `compress_params` are not being used in this function... Do we need it?

**Review by victorusu — CHANGES_REQUESTED** [review]
@vkarak, I liked it. Now, we are using Fortran class syntax (`%`) to describe the class names, right? `ClassName%prefix.parameter_name=parameter_val'`. 😉

**Comment by jjotero** [comment]
These changes in the naming scheme are disruptive and will break tests. If you look at the name produced by the `fullname` method, the test name for a parameterized test without fixtures remains unchanged with respect to previous versions. However, the name produced by `basename` introduces a new notation and may break existing tests if these tests expected the old naming scheme for chaining dependencies and so on.

My second point is that this naming scheme does not solve the problem when the number of parameters is large. Whether you put underscores or `%` doesn't make much difference if you have a large number of parameters, or if some of your parameters are lists, dictionaries, or other structure that would make the name very long. My intention with the `fullname` method is that we eventually move to use `compress_params=True` always, which means that the name complexity will always be _O(1)_ with respect the parameter space (whereas this proposed method is still _O(n)_). With `compress_params=True`, the test would only have the variant id appended to the name, but the user can still query which are the parameter values in that variant through the `get_variant_info` classmethod.

**Comment by vkarak** [comment]
As we discussed today we will use that as a human-readable version of the name, which we could merge later or apply it in a different way when we are listing the tests. We will go with the compressed naming convention, such as `MyTest@<variant_num>` for parametrised tests (normally or through fixtures) and `Fixture@<variant_num>~<fixt_hash>`. The `fixt_hash` captures the scope of the fixture. As @jjotero proposed, the default naming scheme for tests should not change, but we should introduce a configuration parameter to control this, such `use_legacy_naming_scheme`, whose default we'll toggle to true, as soon as we move to 4.0.

**Pull request closed without merging**